File: Sources/ATTNParameterValidation.m
//
//  ATTNParameterValidation.m
//  Example
//
//  Created by Olivia Kim on 11/22/22.
//

#import <Foundation/Foundation.h>
#import "ATTNParameterValidation.h"


@implementation ATTNParameterValidation


+ (bool)isNotNil:(nullable NSObject *)inputValue {
  return ![ATTNParameterValidation isNil:inputValue];
}

+ (bool)isNil:(nullable NSObject *)inputValue {
  return inputValue == nil || [inputValue isKindOfClass:[NSNull class]];
}

+ (bool)isString:(nullable NSObject *)inputValue {
  return [inputValue isKindOfClass:[NSString class]];
}

+ (bool)isEmpty:(nullable NSString *)inputValue {
  return [inputValue length] == 0;
}

+ (bool)isStringAndNotEmpty:(nullable NSObject *)inputValue {
  return ([inputValue isKindOfClass:[NSString class]] && [(NSData *)inputValue length] > 0);
}

+ (void)verifyNotNil:(nonnull NSObject *)inputValue inputName:(nonnull const NSString *)inputName {
  if ([ATTNParameterValidation isNil:inputValue]) {
    [NSException raise:@"Input Was Nil" format:@"%@ should be non-nil", inputName];
  }
}

+ (void)verifyStringOrNil:(nullable NSString *)inputValue inputName:(nonnull const NSString *)inputName {
  if ([ATTNParameterValidation isNotNil:inputValue] && ![ATTNParameterValidation isStringAndNotEmpty:inputValue]) {
      [NSException raise:@"Bad Identifier" format:@"%@ should be a non-empty NSString", inputName];
    }
}

+ (void)verifyString:(nullable NSString *)inputValue inputName:(nonnull const NSString *)inputName {
  if ([ATTNParameterValidation isNil:inputValue]) {
    return;
  }
  if (![ATTNParameterValidation isString:inputValue]) {
    [NSException raise:@"Bad Identifier" format:@"%@ should be NSString", inputName];
  }
  if ([ATTNParameterValidation isEmpty:inputValue]) {
    NSLog(@"Identifier %@ should be non-empty", inputName);
  }
}

+ (void)verify1DStringDictionaryOrNil:(nullable NSDictionary *)inputValue inputName:(nonnull const NSString *)inputName {
  if (![ATTNParameterValidation isNotNil:inputValue])
    return;

  if (![inputValue isKindOfClass:[NSDictionary class]]) {
    [NSException raise:@"Bad Identifier" format:@"%@ should be of form NSDictionary<NSString *, NSString *> *", inputName];
  }

  for (id key in inputValue) {
    [ATTNParameterValidation verifyStringOrNil:[inputValue objectForKey:key] inputName:[NSString stringWithFormat:@"%@[%@]", inputName, key]];
  }
}


@end

File: Sources/ATTNParameterValidation.swift
//
//  ATTNParameterValidation.swift
//  attentive-ios-sdk-framework
//
//  Created by Vladimir - Work on 2024-05-29.
//

import Foundation

struct ATTNParameterValidation {
  static func isNotNil(_ inputValue: NSObject?) -> Bool {
    !isNil(inputValue)
  }

  static func isStringAndNotEmpty(_ inputValue: NSObject?) -> Bool {
    !((inputValue as? String)?.isEmpty ?? true)
  }

  static func verifyNotNil(_ inputValue: NSObject?, inputName: String) {
    guard isNil(inputValue) else { return }
    Loggers.event.debug("Input was nil; \(inputName) should be non-nil")
  }

  static func verifyStringOrNil(_ inputValue: NSObject?, inputName: String) {
    guard isNil(inputValue) else { return }
    guard !isString(inputValue) || isEmpty(inputName) else { return }
    Loggers.event.debug("Identifier \(inputName) should be a non-empty NSString")
  }

  static func verify1DStringDictionaryOrNil(_ inputValue: NSDictionary?, inputName: String) {
    guard isNil(inputValue), let inputValue = inputValue else { return }

    for (key, value) in inputValue {
      verifyStringOrNil(
        value as? NSObject,
        inputName: String(format: "%@[%@]", inputName, key as? String ?? ""))
    }
  }
}

fileprivate extension ATTNParameterValidation {
  static func isNil(_ inputValue: NSObject?) -> Bool {
    inputValue == nil || (inputValue is NSNull)
  }

  static func isString(_ inputValue: NSObject?) -> Bool {
    inputValue is String || inputValue is NSString
  }

  static func isEmpty(_ inputValue: String?) -> Bool {
    inputValue?.isEmpty ?? true
  }
}
